lib/relationship_ancestry_converter_helper.rb

```diff
@@ -0,0 +1,105 @@
+module RelationshipAncestryConverterHelper
+  # src is relationship, dest is vm
+  def transfer_relationships_to_ancestry(dest_model, resource_type, relationship)
+    ancestry_resources = ancestry_resource_ids(relationship, resource_type, dest_model.rails_sequence_range(dest_model.my_region_number))
+    ancestry_sources = ancestry_src_ids(ancestry_resources)
+    new_ancestors = ancestry_of_src_ids_for_src(ancestry_sources)
+    update_src(new_ancestors, dest_model)
+  end
+
+  # src is vm, dest is relationship
+  def create_relationships_from_ancestry(src_model, dest_model, resource_type, relationship)
+    children = src_model.select("ancestry::bigint").where("ancestry not like '%/%'")
+    rels = src_model.where.not(:ancestry => nil).or(src_model.where(:id => children)).map do |obj|
+      dest_model.create!(:relationship  => relationship,
+                         :resource_type => resource_type,
+                         :resource_id   => obj.id,
+                         :ancestry      => obj.ancestry)
+    end.index_by(&:resource_id)
+    rels.each do |_, r|
+      next if r.ancestry.nil?
+
+      ancestry = r.ancestry.split('/').map { |rel| rels[rel.to_i].id }.join('/')
+      r.update!(:ancestry => ancestry)
+    end
+  end
+
+  private
+
+  #
+  # build sql that converts relationship ancestry with the id of the relationship to model-side ancestry with the id of the model
+  #
+
+  # def ancestry_resource_ids(relationship, resource_type, id_range)
+  #   Relationship.where(
+  #     :relationship => relationship,
+  #     :resource_type => resource_type,
+  #     :resource_id => id_range
+  #   ).where.not(:ancestry => nil).flat_map do |a_rels|
+  #     a_rels.ancestry.split('/').each_with_index.map { |rel_id, indx| [a_rels.resource_id, rel_id, indx] }
+  #   end
+  # end
```

Comment:
I'm curious about these commented bits...are they for the other PR in the future?  If so, should they just be added at that time?

Reply:
I wanted ruby versions of the sql for each method just to make it easier to understand.

Reply:
ohhhh I see - I didn't realize they were the same/related method but Ruby - wonder if it's better to organize them together - Something like:

Before:

```ruby
  # def ancestry_resource_ids(relationship, resource_type, id_range)
  #   Relationship.where(
  #     :relationship => relationship,
  #     :resource_type => resource_type,
  #     :resource_id => id_range
  #   ).where.not(:ancestry => nil).flat_map do |a_rels|
  #     a_rels.ancestry.split('/').each_with_index.map { |rel_id, indx| [a_rels.resource_id, rel_id, indx] }
  #   end
  # end

  def ancestry_resource_ids(relationship, resource_type, id_range)
    <<-SQL
    SELECT a_rels.resource_id AS src_id, relationships.id AS rel_id, relationships.indx AS rel_indx
    FROM   relationships a_rels
    LEFT JOIN LATERAL UNNEST(STRING_TO_ARRAY(a_rels.ancestry, '/')::BIGINT[])
    WITH ORDINALITY AS relationships(id, indx) ON TRUE
    WHERE  a_rels.relationship = '#{relationship}'
    AND    a_rels.resource_type = '#{resource_type}'
    AND    a_rels.resource_id BETWEEN #{id_range.first} AND #{id_range.last}
    SQL
  end
```

After:

```ruby
  def ancestry_resource_ids(relationship, resource_type, id_range)
    <<-SQL
    SELECT a_rels.resource_id AS src_id, relationships.id AS rel_id, relationships.indx AS rel_indx
    FROM   relationships a_rels
    LEFT JOIN LATERAL UNNEST(STRING_TO_ARRAY(a_rels.ancestry, '/')::BIGINT[])
    WITH ORDINALITY AS relationships(id, indx) ON TRUE
    WHERE  a_rels.relationship = '#{relationship}'
    AND    a_rels.resource_type = '#{resource_type}'
    AND    a_rels.resource_id BETWEEN #{id_range.first} AND #{id_range.last}
    SQL

    # The above is a pure-SQL version of this similar Ruby code:
    #
    # Relationship.where(
    #   :relationship => relationship,
    #   :resource_type => resource_type,
    #   :resource_id => id_range
    # ).where.not(:ancestry => nil).flat_map do |a_rels|
    #   a_rels.ancestry.split('/').each_with_index.map { |rel_id, indx| [a_rels.resource_id, rel_id, indx] }
    # end
  end
```

Reply:
Even with the redundant method signatures, @Fryguy, you struggled to understand what this is doing, so I don't exactly think taking them out is going to help. This PR's kind of fragile, and this is about the third attempt of this PR.

Reply:
No I think combining them together makes the SQL easier to understand, which was my biggest struggle...I can use the two together to understand the intent.